src/apps/competitions/tests/test_phase_migration.py

```diff
@@ -0,0 +1,98 @@
+        self.normal_user = UserFactory(username='norm')
+        self.normal_user2 = UserFactory(username='norm2')
+        self.normal_user3 = UserFactory(username='norm3')
+        self.competition = CompetitionFactory(created_by=self.owner)
+        self.competition_participant = CompetitionParticipantFactory(user=self.normal_user,
+                                                                     competition=self.competition)
+        self.competition_participant2 = CompetitionParticipantFactory(user=self.normal_user2,
+                                                                      competition=self.competition)
+        self.competition_participant3 = CompetitionParticipantFactory(user=self.normal_user3,
+                                                                      competition=self.competition)
+        self.phase1 = PhaseFactory(competition=self.competition, auto_migration=True, status='Current')
+        self.phase2 = PhaseFactory(competition=self.competition, auto_migration=True, status='Next')
+        self.phase3 = PhaseFactory(competition=self.competition, auto_migration=False, status='Final')
+        self.submission = self.make_submission()
+        self.submission2 = self.make_submission(participant=self.competition_participant2)
+        self.client = APIClient()
+
+    def make_submission(self, **kwargs):
+        kwargs.setdefault('owner', self.owner)
+        kwargs.setdefault('participant', self.competition_participant)
+        kwargs.setdefault('phase', self.phase1)
+        kwargs.setdefault('status', 'None')
+        sub = SubmissionFactory(**kwargs)
+        SubmissionScoreFactory(submission=sub)
+        return sub
+
+    def test_migrate_submissions(self):
+        """
+        Determine whether all submissions in phase1 end up in phase2
+        :return:
+        """
+        assert self.phase2.submissions.count() == 0
+        self.competition.apply_phase_migration(self.phase1, self.phase2)
+        assert self.phase1.submissions.count() == self.phase2.submissions.count()
+
+    def test_migrating_competitions(self):
+        """
+        Make sure competitions which are already migrating do not migrate.
+        :return:
+        """
+        self.competition.is_migrating = True
+        self.competition.apply_phase_migration(self.phase1, self.phase2)
+        self.assertFalse(self.phase1.submissions.count() == self.phase2.submissions.count())
+
+    def test_migrating_scoring_submissions(self):
+        """
+        Test to see if there are submissions being scored currently. If there are, delay migration.
+        :return:
+        """
+        self.submission3 = self.make_submission(status='Scoring', participant=self.competition_participant3)
+        self.competition.apply_phase_migration(self.phase1, self.phase2)
+        self.assertFalse(self.phase1.submissions.count() == self.phase2.submissions.count())
+
+    def test_auto_migrate(self):
```

Comment:
We should make another test that makes sure we don't auto migrate if the phase isn't supposed to auto migrate. Like maybe modify phase 2 to `auto_migrate=False` for a test and make sure we migrate from 1 -> 2 but not 2 -> 3.

Reply:
Named variables more clearly by naming it  `auto_migrate_to_this_phase`. Also added a test where we try to push to an `auto_migrate_to_this_phase = False` in phase 3 after `auto_migrate_to_this_phase = True` in phase 2.